pkg/apis/devopsconsole-operator/v1alpha1/gitsource_types.go

```diff
@@ -0,0 +1,70 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// GitSourceSpec defines the desired state of GitSource
+// +k8s:openapi-gen=true
+type GitSourceSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
+	// Add custom validation using kubebuilder tags: https://book.kubebuilder.io/beyond_basics/generating_crd.html
+
+	// URL of the git repo
+	URL string `json:"url"`
+
+	// Ref is a git reference. Optional. "master" is used by default.
+	Ref string `json:"ref,omitempty"`
+
+	// ContextDir is a path to subfolder in the repo. Optional.
+	ContextDir string `json:"contextDir,omitempty"`
+
+	// HttpProxy is optional.
+	HttpProxy string `json:"httpProxy,omitempty"`
+
+	// HttpsProxy is optional.
+	HttpsProxy string `json:"httpsProxy,omitempty"`
+
+	// NoProxy can be used to specify domains for which no proxying should be performed. Optional.
+	NoProxy string `json:"noProxy,omitempty"`
+
+	// SourceSecret is the name of the secret that contains credentials to access the git repo. Optional.
+	SourceSecret string `json:"sourceSecret,omitempty"`
+}
+
+// GitSourceStatus defines the observed state of GitSource
+// +k8s:openapi-gen=true
+type GitSourceStatus struct {
```

Comment:
Does it make sense to maybe store the last fetched/seen commit hash? I mean, if the `ref` is `master` it might still be interesting to store the commit hash that `master` resolved to the last time this repo was seen.

Reply:
Do you have any examples of how we could use such info?

Reply:
When you have a git source configured and you wonder why your build still doesn't contain a fix it might be useful to know of the source is at the correct hash.

Reply:
Ok I see your point but I’m not sure if we should keep this info here or somewhere else. Let’s introduce it later when we need it.

Reply:
I'm not arguing with that possible use case. I just don't think we really have to introduce this now. It's not going to be used anytime soon. I'm not sure if it's really the best place to be. So, it's premature for now. We can considering adding this field later when we are ready to support tracking commits.

Reply:
Having the hash stored somewhere looks very interesting.
Hi @kwk, Is this the kind of flow you have in mind?

- A gitSource CR is created and a component CR is created ( which references the gitSource CR ).
- The controller listening on the creation of a new  component CR, creates the buildConfig ( and with it a lot of other things ).
- The 'watch' loop gets to know that a buildConfig *belonging* to the component was created. That could write some status regarding the buildConfig to the `component` CR.
- When a new `build` is created, there's a block which says:
```
revision:
    git:
      author:
        email: sbose78@gmail.com
        name: Shoubhik Bose
      commit: a4743fd8644577b4a0457a552825cb4f1bdd7f3a
      committer:
        email: noreply@github.com
        name: GitHub
      message: parameterize db name
    type: Git
```

- So, if we 'listen' on "Builds in namespace X, not necessarily owned by component C", then we could get this commit information and add it to the `component` status.

Is that what you are saying?